sled-agent/config-reconciler/src/ledger.rs

```diff
@@ -182,6 +246,914 @@ struct LedgerTask<T> {
+        // anything, and `wait_for_m2_disks` should not return until we've seen
+        // one.
+        if let Err(err) = self.wait_for_m2_disks().await {
+            return Err((self.log, err));
+        }
+        assert_ne!(
+            *self.current_config_tx.borrow(),
+            CurrentSledConfig::WaitingForInternalDisks
+        );
+
+        loop {
+            let Some(request) = self.request_rx.recv().await else {
+                return Err((
+                    self.log,
+                    LedgerTaskExit::LedgerTaskHandleDropped,
+                ));
+            };
+
+            match request {
+                LedgerTaskRequest::WriteNewConfig { new_config, tx } => {
+                    // We don't care if the receiver is gone.
+                    _ = tx.send(self.set_new_config(new_config).await);
+                }
+                LedgerTaskRequest::ValidateArtifactConfig {
+                    new_config,
+                    tx,
+                } => {
+                    // We don't care if the receiver is gone.
+                    _ = tx.send(self.validate_artifact_config(new_config));
+                }
+            }
+        }
+    }
+
+    async fn set_new_config(
+        &mut self,
+        new_config: OmicronSledConfig,
+    ) -> Result<(), LedgerNewConfigError> {
+        let internal_disks = self.internal_disks_rx.current_and_update();
+        let mut config_datasets =
+            internal_disks.all_config_datasets().peekable();
+
+        // This would be very unusual: We waited for at least one M.2 to be
+        // present in `run_impl`, so if this is empty that means we've lost all
+        // M.2s. Refuse to accept new config.
+        if config_datasets.peek().is_none() {
+            error!(self.log, "no internal disks available any longer?!");
+            return Err(LedgerNewConfigError::NoInternalDisks);
+        }
+
+        // Check that we can accept the new ledger.
+        self.validate_sled_config(&new_config).await?;
+
+        // `validate_sled_config` confirmed that the new config is valid in
+        // terms of generation numbers. The most common valid case is
+        // "generation the same, contents unchanged", because Nexus periodically
+        // sends us our current config regardless of whether we already have it.
+        // In that most common case, we now fall through to committing the
+        // ledger to disk. This seems unnecessary: if the generation matches, we
+        // already committed the ledger previously. But we don't know that it's
+        // been written to both M.2s:
+        //
+        // * It's possible only one M.2 was available when we previously
+        //   committed
+        // * It's possible we had two M.2s, but one of the writes failed
+        //   (`Ledger::commit()` returns `Ok(())` as long as at least one of the
+        //   writes succeeds)
+        //
+        // We could probably address both of those and avoid spurious writes,
```

Comment:
This is exactly the same choice we made with clickhouse configuration, except not because of having 2 M.2s to write to. We could end up in partially written state in some cases (multiple files), and instead of doing an epoch check we just rewrite every time. Built in resilience!